app/models/game.rb

```diff
@@ -70,17 +75,33 @@ def forfeit(current_user)
   end
 
   def stalemate?(is_white)
-    return false if check?(is_white)
+    return false if check?
+    king = pieces.find_by(is_white: is_white, type: KING)
     (1..8).each do |new_x|
       (1..8).each do |new_y|
-        pieces.where(is_white: is_white).where.not(x_position: nil, y_position: nil).find_each do |piece|
+        pieces.where(is_white: is_white).where.not(x_position: nil, y_position: nil, type: KING).find_each do |piece|
           return false if piece.legal_move?(new_x, new_y)
         end
+        return true if king.legal_move?(new_x, new_y) && under_attack?(is_white, new_x, new_y)
       end
     end
     true
   end
 
+  # Use this code to test the stalemate rspec tests:
+  # (1..8).each do |new_x|
+  #   (1..8).each do |new_y|
+  #     pieces.where(is_white: true).where.not(x_position: nil, y_position: nil, type: KING).find_each do |piece|
+  #       puts "#{[new_x, new_y]}: legal move for #{piece.type} at [#{piece.x_position}, #{piece.y_position}]? #{piece.legal_move?(new_x, new_y)}"
+  #     end
+  #   end
+  # end
+
+  def attacking_piece
```

Comment:
This logic is faulty. A piece that moved last could have simply exposed the king to check, but could be different from the piece putting the king's square under attack. How else can we do this?

Reply:
Instead of finding last piece the attacking player moved, I would cycle through each of that player's pieces to find first one that currently has the king in check. Doesn't matter if there are multiple pieces currently holding king in check, if the player with king in check can't escape or eliminate the first one, we don't need to look at others.